### ISSUE TITLE
Add `withArgs()` to validate via closure passed arguments at once

### ISSUE THREAD

**Issue by gmazzap** [mockery/mockery#459]
I think would be handy a method like

``` php
withArgs($closure)
```

that would work similar to 

``` php
with( Mockery::on($closure) );
```

with the only difference that in first case the `$closure`receives **all** the arguments passed to the target method.

**Comment by aik099** [comment]
With proposed `withArgs($closure)` syntax it's impossible to uniquely detect which expectation should fire because one with close will be used in all cases.

Instead I recommend using `->andReturnUsing($closure)`. Then the `$closure` receives all the arguments passed.

**Comment by Giuseppe-Mazzapica** [comment]
Thanks @aik099 for your response.

To use `andReturnUsing` was my first thought, but to make test fail if arguments are not valid, the only way is to manually throw an exception, and it's quite verbose because an explicative message should be added too or is hard to understand why the test failed.

The only aim of my proposal is to have less verbose test code, so...

Well, yes, I assume that `withArgs` should be used as unique or last expectation for a method.

Thats IMO should not be a problem: for example, currently you can't mix `andReturnUsing` and `andReturn` in same test for same method, and in the same way you couldn't mix `with` and `withArgs`...

**Comment by beni0888** [comment]
@padraic @aik099 @Giuseppe-Mazzapica So, what do you think about this feature? Should it be added to the API? Can I start working on it?

**Comment by davedevelopment** [comment]
Can for me!

**Comment by aik099** [comment]
How we can distinguish between 2 cases:

* when `withArgs($closure)` means a `I want this method to be called with 1 argument, which is closure`
* when `withArgs($closure)` means a `I want the closure to handle argument matching for me`
?

**Comment by aik099** [comment]
Ah, the `withArgs` get's called inside `with` using `func_get_args()`. In that case, yes it's possible to make `withArgs` to accept a closure.

**Comment by davedevelopment** [comment]
Yeah, `withArgs` as it stands is intended to take an array of all expected arguments, so we can just switch on type.

**Comment by beni0888** [comment]
ok, I'll try it :)

**Comment by peixinchen** [comment]
In my view, withArgs($argsOr$cloure) should be separated as withArgs(array $args) and withArgsUsing(Closure $closure), and it could solve the problom.

**Comment by Revisor** [comment]
Yeah, just came here to say the same. The documentation for 0.9 regarding `withArgs()` doesn't match the code.

**Comment by davedevelopment** [comment]
I'm not sure I understand, didn't this get merged in to the master branch? If someone could point me directly to the differences, I'll take a look.

**Comment by Revisor** [comment]
[The documentation](http://docs.mockery.io/en/latest/reference/expectations.html), whose title says "Mockery Docs 0.9 documentation" contains this:

> ```withArgs(closure)```
> 
> Instead of providing a built-in matcher for each argument, you can provide a closure that matches all passed arguments at once. The given closure receives all the arguments passed in the call to the expected method. In this way, this expectation only applies to method calls where passed arguments make the closure evaluates to true.

However that is not true for Mockery 0.9, where `Expectation::withArgs()` [expects an array.](https://github.com/padraic/mockery/blob/0.9/library/Mockery/Expectation.php#L382)

It is also not true for the master branch, where `withArgs()` [expects either an array (like before), or a closure.](https://github.com/padraic/mockery/blob/master/library/Mockery/Expectation.php#L404)

**Comment by davedevelopment** [comment]
@Revisor Thanks! I think there are a few problems in play here, will try and get them sorted.

**Comment by davedevelopment** [comment]
Fixed the documentation title in 0f183fa38026cf4779

So now the [stable](http://docs.mockery.io/en/stable/) docs have "Mockery docs 0.9" in the title, the [latest](http://docs.mockery.io/en/latest/) have "Mockery docs 1.0-alpha". 

The documentation for the `withArgs([$arg1, $arg2])` syntax is just above the `withArgs($closure)` documentation, but it's not all that clear, perhaps it should be a separate code block.

**Comment by Revisor** [comment]
Thank you.
Ah, I didn't notice `withArgs` in the block above, my eyes skimmed it.

May I suggest, instead of
```
with(arg1, arg2, ...) / withArgs(array(arg1, arg2, ...))
```
Format it like this
```
with(arg1, arg2, ...)
withArgs(array(arg1, arg2, ...))
```
What do you think?

**Comment by davedevelopment** [comment]
Sure, done in 89f7708

**Comment by halfer** [comment]
@Revisor: if I want to use a closure to match method params, do I have to use the master branch at present? The change in master to support closures went in on 1 March 2016 (b4b978ca75) and it does not seem to be in 0.9.6 or 0.9.7.

**Comment by Revisor** [comment]
That's more a question for davedevelopment, but as far as I can see, there is currently no tagged version supporting closure as an argument for `withArgs()`. You would have to surf dangerously on the development branch.

But maybe Dave could tag the current version as 1.0 alpha?

**Comment by halfer** [comment]
Thanks @Revisor, well surfing dangerously it is then `:)`. I will revisit in a month or two to see if a new release has been made.

**Comment by rparsi-commer** [comment]
@davedevelopment @Revisor @halfer 
What's the update on this?
I need to be able to match method params using a closure, to tell Mockery
if the params passed to the mocked method are for the specific expectation.

**Comment by halfer** [comment]
@rparsi-commer: I had to repair my Composer references recently. The latest release 0.9.9 does not have the closure facility, so that still requires `dev-master`. Unfortunately that also bumped up the minimum PHP version - I am still using Ubuntu 14.04 for development, so was still using 5.5. It gave me a good excuse to upgrade to 7.x, which is pretty easy to do.

It would be nice to see a release tag including this, but if you are stuck and don't want to build against HEAD, fork the code and create your own tag. You can point your `composer.json` at that instead.

**Comment by rparsi-commer** [comment]
@halfer So closures don't work at all?
I'm using Mockery version 0.9.x-dev (branch alias), I think my actual installed version is older than 0.9.4.
I have also tried using Expectation::andReturnUsing($closure).

No matter what I do my closure doesn't get called.
I've also tried Expectation::ordered() when creating the expectations.

I'm mocking a method that has signature ($stringValue, BlahInterface $blah) where
the actual value of properties in $blah differs on each call.

How do I do this with current stable release?

**Comment by halfer** [comment]
> So closures don't work at all?

No @rparsi-commer, they work just fine. We're probably talking at cross-purposes; my point was that there is not a release for this, so you have to build against the master of Mockery.

Here is the line in my `composer.json`:

        "mockery/mockery": "dev-master",

> How do I do this with current stable release?

You don't, it is not in a stable release. As I have explained, if you don't want to point at a moving target, fork the code yourself and point at that.

**Comment by rparsi-commer** [comment]
@halfer Ok, thanks for the clarification.

**Comment by davedevelopment** [comment]
There is a 1.0.0-alpha1 tag, I'd recommend moving to that. I find it stable enough for my tests.

**Comment by halfer** [comment]
Lovely, will do @davedevelopment.

**Comment by robertbasic** [comment]
If I'm understanding this thread correctly, the feature requested is implemented, and available since 1.0.0-alpha1 tag, so I'm going to close this issue.

Please open a new issue if there's still something amiss with this.
